engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskCompleteProcessor.java

```diff
@@ -0,0 +1,92 @@
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedCommandWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedRejectionWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.engine.state.immutable.ElementInstanceState;
+import io.camunda.zeebe.engine.state.immutable.ProcessingState;
+import io.camunda.zeebe.engine.state.immutable.UserTaskState;
+import io.camunda.zeebe.engine.state.immutable.UserTaskState.LifecycleState;
+import io.camunda.zeebe.engine.state.instance.ElementInstance;
+import io.camunda.zeebe.protocol.impl.record.value.usertask.UserTaskRecord;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.protocol.record.intent.UserTaskIntent;
+import io.camunda.zeebe.stream.api.records.TypedRecord;
+import java.util.List;
+
+public final class UserTaskCompleteProcessor implements TypedRecordProcessor<UserTaskRecord> {
+
+  private final UserTaskState userTaskState;
+  private final ElementInstanceState elementInstanceState;
+  private final EventHandle eventHandle;
+  private final StateWriter stateWriter;
+  private final TypedCommandWriter commandWriter;
+  private final TypedRejectionWriter rejectionWriter;
+  private final UserTaskCommandPreconditionChecker preconditionChecker;
+
+  public UserTaskCompleteProcessor(
+      final ProcessingState state, final EventHandle eventHandle, final Writers writers) {
+    userTaskState = state.getUserTaskState();
+    elementInstanceState = state.getElementInstanceState();
+    this.eventHandle = eventHandle;
+    stateWriter = writers.state();
+    commandWriter = writers.command();
+    rejectionWriter = writers.rejection();
+    preconditionChecker =
+        new UserTaskCommandPreconditionChecker(
+            List.of(LifecycleState.CREATED), "complete", userTaskState);
+  }
+
+  @Override
+  public void processRecord(final TypedRecord<UserTaskRecord> userTaskRecord) {
+    preconditionChecker
+        .check(userTaskRecord)
+        .ifRightOrLeft(
+            ok -> {
+              final long userTaskKey = userTaskRecord.getKey();
+              final UserTaskRecord persistedUserTask =
+                  userTaskState.getUserTask(userTaskKey, userTaskRecord.getAuthorizations());
+
+              persistedUserTask.setVariables(userTaskRecord.getValue().getVariablesBuffer());
+
+              stateWriter.appendFollowUpEvent(
+                  userTaskKey, UserTaskIntent.COMPLETING, persistedUserTask);
+              stateWriter.appendFollowUpEvent(
+                  userTaskKey, UserTaskIntent.COMPLETED, persistedUserTask);
+              completeElementInstance(persistedUserTask);
+            },
+            violation ->
+                rejectionWriter.appendRejection(
+                    userTaskRecord, violation.getLeft(), violation.getRight()));
```

Comment:
:x: You're also missing `responseWriter` statements when a job is completed and rejected, similar to how it's done [here](https://github.com/camunda/zeebe/blob/2e22958504385ec243ffd4e932c2283b3ad5104f/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java#L104) and [here](https://github.com/camunda/zeebe/blob/2e22958504385ec243ffd4e932c2283b3ad5104f/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java#L124).

With the `rejectionWriter` and `stateWriter` calls we append events/rejections on the log stream, but we don't provide any response for the client.

Reply:
Done with https://github.com/camunda/zeebe/pull/15471/commits/58da3e40bbff74c5fc298092a40018f9167f9834

Reply:
> 💭 Overall, I feel like there is a lot of implicit knowledge required to write a command processor (e.g. you need to know which writers to invoke). I would really like if this would be more explicit, e.g. by returning a CommandResult from the processRecord method that contains the events, commands, and response or the rejection. Something like that. Right now, you need to discover the different aspects by looking at different existing commands to discover this. That's making it harder for external contributors as well, I guess.

@tmetzke I agree. We need to improve the development process for command processors. Maybe we can create an issue for this and list all the issues that we know of, for example:
* Remove the `CommandProcessor` interface, to avoid its usage.
* Document how to decide what `writers` will be needed. Currently it's not clear, and each dev needs to discover patterns from previous processor implementations. Event these aren't consistent since even empty (gRPC) responses use the `responseWriter`, so it's not clear when we need it, and when we don't.

Reply:
Added this to #8003